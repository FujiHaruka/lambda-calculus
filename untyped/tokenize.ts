import { TokenizerContext } from "./tokenizers/TokenizerContext.ts";
import { Token } from "./tokenizers/types.ts";
import { scanSymbol } from "./tokenizers/scanSymbol.ts";
import { scanVar } from "./tokenizers/scanVar.ts";
import { UnexpectedCharacterError } from "./tokenizers/errors.ts";

function scan(ctx: TokenizerContext): Token | null {
  return scanSymbol(ctx) || scanVar(ctx);
}

/**
 * Tokenize the given code.
 */
export function tokenize(code: string): Token[] {
  const ctx = new TokenizerContext({ code, position: 0 });

  ctx.skipSpaces();
  while (!ctx.isEOF) {
    const maybeToken = scan(ctx);
    if (maybeToken) {
      ctx.pushToken(maybeToken);
    } else {
      throw new UnexpectedCharacterError(ctx);
    }
    ctx.skipSpaces();
  }

  return ctx.tokens;
}
