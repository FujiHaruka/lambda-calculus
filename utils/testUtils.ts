export {
  assertEquals,
  assertStrictEquals,
  assertStringIncludes,
  assertThrows,
} from "https://deno.land/std@0.192.0/testing/asserts.ts";
export { describe, it } from "https://deno.land/std@0.192.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.192.0/testing/snapshot.ts";
export { walkSync } from "https://deno.land/std@0.192.0/fs/mod.ts";
export * as path from "https://deno.land/std@0.192.0/path/mod.ts";
