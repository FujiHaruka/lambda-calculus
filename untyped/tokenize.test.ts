import { assertSnapshot, it } from "./testUtils.ts";
import { tokenize } from "./tokenize.ts";

[
  "",
  " ",
  "x",
  " xyz ",
  "x->x",
  "x x",
  "x -> y -> z",
  "xxx_000 -> yyy_111 -> zzz_222",
  "( x  ->  y )  ( x  ->  y )",
].forEach((code) => {
  it(`tokenizes "${code}"`, async (t) => {
    await assertSnapshot(t, tokenize(code));
  });
});

it("tokenizes with eof token", async (t) => {
  await assertSnapshot(t, tokenize("x", { eofToken: true }));
})
