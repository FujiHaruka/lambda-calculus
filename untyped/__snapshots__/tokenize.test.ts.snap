export const snapshot = {};

snapshot[`tokenize "" 1`] = `[]`;

snapshot[`tokenize " " 1`] = `[]`;

snapshot[`tokenize "x" 1`] = `
[
  {
    end: 1,
    start: 0,
    type: "var",
  },
]
`;

snapshot[`tokenize " xyz " 1`] = `
[
  {
    end: 4,
    start: 1,
    type: "var",
  },
]
`;

snapshot[`tokenize "x->x" 1`] = `
[
  {
    end: 1,
    start: 0,
    type: "var",
  },
  {
    end: 3,
    start: 1,
    type: "arrow",
  },
  {
    end: 4,
    start: 3,
    type: "var",
  },
]
`;

snapshot[`tokenize "x x" 1`] = `
[
  {
    end: 1,
    start: 0,
    type: "var",
  },
  {
    end: 3,
    start: 2,
    type: "var",
  },
]
`;

snapshot[`tokenize "x -> y -> z" 1`] = `
[
  {
    end: 1,
    start: 0,
    type: "var",
  },
  {
    end: 4,
    start: 2,
    type: "arrow",
  },
  {
    end: 6,
    start: 5,
    type: "var",
  },
  {
    end: 9,
    start: 7,
    type: "arrow",
  },
  {
    end: 11,
    start: 10,
    type: "var",
  },
]
`;

snapshot[`tokenize "xxx_000 -> yyy_111 -> zzz_222" 1`] = `
[
  {
    end: 7,
    start: 0,
    type: "var",
  },
  {
    end: 10,
    start: 8,
    type: "arrow",
  },
  {
    end: 18,
    start: 11,
    type: "var",
  },
  {
    end: 21,
    start: 19,
    type: "arrow",
  },
  {
    end: 29,
    start: 22,
    type: "var",
  },
]
`;

snapshot[`tokenize "( x  ->  y )  ( x  ->  y )" 1`] = `
[
  {
    end: 1,
    start: 0,
    type: "left_paren",
  },
  {
    end: 3,
    start: 2,
    type: "var",
  },
  {
    end: 7,
    start: 5,
    type: "arrow",
  },
  {
    end: 10,
    start: 9,
    type: "var",
  },
  {
    end: 12,
    start: 11,
    type: "right_paren",
  },
  {
    end: 15,
    start: 14,
    type: "left_paren",
  },
  {
    end: 17,
    start: 16,
    type: "var",
  },
  {
    end: 21,
    start: 19,
    type: "arrow",
  },
  {
    end: 24,
    start: 23,
    type: "var",
  },
  {
    end: 26,
    start: 25,
    type: "right_paren",
  },
]
`;
